lib/chef/resource/chef_client_config.rb

```diff
@@ -0,0 +1,307 @@
+        description: %q(An array of hashes that contain a report handler class and the arguments to pass to that class on initialization. The hash should include `class` and `argument` keys where `class` is a String and `argument` is an array of quoted String values. For example: `[{'class' => 'MyHandler', %w('"argument1"', '"argument2"')}]`),
+        default: []
+
+      property :exception_handlers, Array,
+        description: %q(An array of hashes that contain a exception handler class and the arguments to pass to that class on initialization. The hash should include `class` and `argument` keys where `class` is a String and `argument` is an array of quoted String values. For example: `[{'class' => 'MyHandler', %w('"argument1"', '"argument2"')}]`),
+        default: []
+
+      property :chef_license, String,
+        description: "Accept the [Chef EULA](https://www.chef.io/end-user-license-agreement/)",
+        equal_to: %w{accept accept-no-persist accept-silent}
+
+      property :policy_name, String,
+        description: "The name of a policy, as identified by the `name` setting in a Policyfile.rb file. `policy_group`  when setting this property."
+
+      property :policy_group, String,
+        description: "The name of a `policy group` that exists on the #{ChefUtils::Dist::Server::PRODUCT}. `policy_name` must also be specified when setting this property."
+
+      property :named_run_list, String,
+        description: "A specific named runlist defined in the node's applied Policyfile, which the should be used when running #{ChefUtils::Dist::Infra::PRODUCT}."
+
+      property :pid_file, String,
+        description: "The location in which a process identification number (pid) is saved. An executable, when started as a daemon, writes the pid to the specified file. "
+
+      property :file_cache_path, String,
+        description: "The location in which cookbooks (and other transient data) files are stored when they are synchronized. This value can also be used in recipes to download files with the `remote_file` resource."
+
+      property :file_backup_path, String,
+        description: "The location in which backup files are stored. If this value is empty, backup files are stored in the directory of the target file"
+
+      property :run_path, String,
+        description: "The location where #{ChefUtils::Dist::Infra::PRODUCT} should write the PID file."
+
+      property :file_staging_uses_destdir, String,
+        description: "How file staging (via temporary files) is done. When `true`, temporary files are created in the directory in which files will reside. When `false`, temporary files are created under `ENV['TMP']`"
+
+      property :additional_config, String,
+        description: "Additional text to add at the bottom of the client.rb config. This can be used to run custom Ruby or to add less common config options"
+
+      action :create do
+        unless ::Dir.exist?(new_resource.config_directory)
```

Comment:
why the extra idempotency check here and below?

Reply:
They're the equivalent of "create_if_missing" which we don't have for directory. People have historically sharpened pitch forks over the details of creating these directories so this way if they want to do it themselves they do it and this resource won't touch them. It creates the dirs if they aren't there, but it's not actively managing them since it's concerned with the client.rb only. The same thing is done in the other chef_client_* resources.